File: sample-bots/java/Fire/src/main/java/Fire.java
import dev.robocode.tankroyale.botapi.*;
import dev.robocode.tankroyale.botapi.events.*;

import java.io.IOException;

// ------------------------------------------------------------------
// Fire
// ------------------------------------------------------------------
// A sample bot original made for Robocode by Mathew Nelson.
// Ported to Robocode Tank Royale by Flemming N. Larsen.
//
// This bot moves to a corner, then swings the gun back and forth.
// If it dies, it tries a new corner in the next round.
// ------------------------------------------------------------------
public class Fire extends Bot {

    int dist = 50; // Distance to move when we're hit, forward or back
    boolean isScanning; // Flag indicating if onScannedBot() handler is running

    // The main method starts our bot
    public static void main(String[] args) {
        new Fire().start();
    }

    // Constructor, which loads the bot settings file
    Fire() {
        super(BotInfo.fromFile("Fire.json"));
    }

    // Called when a new round is started -> initialize and do some movement
    @Override
    public void run() {
        isScanning = false; // Clear scanning flag for each new turn

        // Set colors
        setBodyColor("#FA0"); // orange
        setGunColor("#F70"); // dark orange
        setTurretColor("#F70"); // dark orange
        setRadarColor("#F00"); // red
        setScanColor("#F00"); // red
        setBulletColor("#08F"); // light blue

        // Spin the gun around slowly... forever
        while (isRunning()) {
            if (isScanning) {
                // Skip a turn if the onScannedBot handler is running
                go();
            } else {
                // Turn the gun a bit if the bot if the target speed is 0
                turnGunLeft(5);
            }
        }
    }

    // We scanned another bot -> fire!
    @Override
    public void onScannedBot(ScannedBotEvent e) {
        isScanning = true; // We are now scanning

        // If the other robot is close by, and we have plenty of life, fire hard!
        double distance = distanceTo(e.getX(), e.getY());
        if (distance < 50 && getEnergy() > 50) {
            fire(3);
        } else {
            // Otherwise, only fire 1
            fire(1);
        }
        // Rescan
        scan();

        isScanning = false; // We are not scanning any more
    }

    // We were hit by a bullet -> turn perpendicular to the bullet, and move a bit
    @Override
    public void onHitByBullet(BulletHitBotEvent e) {
        // Turn perpendicular to the bullet direction
        turnLeft(normalizeRelativeAngle(90 - (getDirection() - e.getBullet().getDirection())));

        // Move forward or backward depending if the distance is positive or negative
        forward(dist);
        dist *= -1; // Change distance, meaning forward or backward direction

        // Rescan
        scan();
    }

    // We have hit another bot -> aim at it and fire hard!
    @Override
    public void onHitBot(HitBotEvent e) {
        // Turn gun to the bullet direction
        double direction = directionTo(e.getX(), e.getY());
        double gunBearing = normalizeRelativeAngle(direction - getGunDirection());
        turnGunLeft(gunBearing);

        // Fire hard
        fire(3);
    }
}

File: sample-bots/python/Fire/Fire.py
from robocode_tank_royale.bot_api.bot import Bot
from robocode_tank_royale.bot_api.color import Color
from robocode_tank_royale.bot_api.events import (
    ScannedBotEvent,
    HitByBulletEvent,
    HitBotEvent,
)


# ------------------------------------------------------------------
# Fire
# ------------------------------------------------------------------
# A sample bot originally made for Robocode by Mathew Nelson.
#
# Sits still, continuously rotates its gun, and only moves when hit.
# ------------------------------------------------------------------
class Fire(Bot):
    """Sits still, rotates its gun, and moves reactively when hit."""

    def __init__(self) -> None:
        super().__init__()
        self._dist: int = 50  # Distance to move when we're hit, forward or back

    def run(self) -> None:
        # Set colors
        self.body_color = Color.from_rgb(0xFF, 0xAA, 0x00)   # orange
        self.turret_color = Color.from_rgb(0xFF, 0x77, 0x00) # dark orange
        self.radar_color = Color.from_rgb(0xFF, 0x00, 0x00)  # red
        self.scan_color = Color.from_rgb(0xFF, 0x00, 0x00)   # red
        self.bullet_color = Color.from_rgb(0x00, 0x88, 0xFF) # light blue

        # Spin the gun around slowly... forever
        while self.running:
            self.turn_gun_right(5)

    def on_scanned_bot(self, e: ScannedBotEvent) -> None:
        # If the other bot is close by, and we have plenty of life, fire hard!
        distance = self.distance_to(float(e.x), float(e.y))
        if distance < 50 and self.energy > 50:
            self.fire(3)
        else:
            # Otherwise, only fire 1
            self.fire(1)
        # Rescan
        self.rescan()

    def on_hit_by_bullet(self, e: HitByBulletEvent) -> None:
        # Turn perpendicular to the bullet direction
        bearing = self.calc_bearing(e.bullet.direction)
        self.turn_right(90 - bearing)

        # Move forward or backward depending on current sign
        self.forward(self._dist)
        self._dist *= -1  # Change sign -> toggles forward/backward

        # Rescan
        self.rescan()

    def on_hit_bot(self, e: HitBotEvent) -> None:
        # Turn gun to face the target
        gun_bearing = self.gun_bearing_to(float(e.x), float(e.y))
        self.turn_gun_right(gun_bearing)
        # Fire hard
        self.fire(3)


def main() -> None:
    bot = Fire()
    bot.start()


if __name__ == "__main__":
    main()
